runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2020-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.grpc.internal
+
+import akka.NotUsed
+import akka.annotation.InternalApi
+import akka.stream.scaladsl.Flow
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+import akka.util.ByteString
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object DecodeBase64 {
+  def apply(): Flow[ByteString, ByteString, NotUsed] =
+    Flow[ByteString].via(new DecodeBase64)
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class DecodeBase64 extends GraphStage[FlowShape[ByteString, ByteString]] {
+  private val in = Inlet[ByteString]("DecodeBase64.in")
+  private val out = Outlet[ByteString]("DecodeBase64.out")
+
+  override def initialAttributes = Attributes.name("DecodeBase64")
+
+  final override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var buffer: ByteString = ByteString.empty
+
+      override def onPush(): Unit = {
+        buffer ++= grab(in)
+
+        val length = buffer.length
+        val decodeLength = length - length % 4
+
+        if (decodeLength > 0) {
+          val (decodeBytes, remaining) = buffer.splitAt(decodeLength)
+          push(out, decodeBytes.decodeBase64)
+          buffer = remaining
+        } else {
+          pull(in)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (buffer.nonEmpty) {
+          if (isAvailable(out)) {
+            push(out, buffer.decodeBase64)
```

Comment:
Can we be sure that the right number of bytes have been pushed to the buffer, or is there a risk of some premature upstream finish? Maybe best to have the same `val decodeLength = length - length % 4` logic here?

Reply:
and same in the push from onPull

Reply:
Padding in base64 is only required for correctness when concatenating encoded chunks. Most decoders will implicitly add padding for any missing bytes at the end of encoded data. I think this is an instance of the robustness principle ("be conservative in what you send, liberal in what you accept").

akka.util.ByteString.decodeBase64 automatically adds trailing padding:
```
scala> val data = akka.util.ByteString(List(0.toByte))
val data: akka.util.ByteString = ByteString(0)

scala> val encoded = data.encodeBase64
val encoded: akka.util.ByteString = ByteString(65, 65, 61, 61)

scala> val decoded = encoded.dropRight(2).decodeBase64
val decoded: akka.util.ByteString = ByteString(0)
```

I followed this same principle when implementing the DecodeBase64 flow. If the stream ends with an unaligned chunk of data, we decode it and pass it downstream before completing the stage. This should allow the akka-grpc servers to work with grpc-web client implementations that omit padding, without affecting correctness.

Reply:
I'm not sure we talk about the same problem. Here in onUpstreamFinish it tries to `decodeBase64` whatever is currently in the buffer. For example, `ByteString.fromString("a")` but only 1 byte has been pushed to the buffer and then the stream is finished. That can probably not happen for well behaved encoding upstreams but should this flow fail in such case or should it ignore the incomplete trailing part?

```
scala> ByteString.fromString("a").encodeBase64.take(1).decodeBase64.toString
java.lang.IllegalArgumentException: Input byte[] should at least have 2 bytes for base64 bytes
  at java.base/java.util.Base64$Decoder.decodedOutLength(Base64.java:709)
  at java.base/java.util.Base64$Decoder.decode(Base64.java:662)
  at akka.util.ByteString$ByteString1.decodeBase64(ByteString.scala:384)
```

Reply:
> That can probably not happen for well behaved encoding upstreams but should this flow fail in such case or should it ignore the incomplete trailing part?

I don't think we should silently ignore/drop unaligned input.

For unaligned-but-parsable input (`ByteString.fromString("a").encodeBase64.take(2)`) I'm OK with either parsing it or throwing an error. The current implementation parses it, which seems fine by me.

For unaligned unparsable input (`ByteString.fromString("a").encodeBase64.take(1)`) we should fail the stream. Here we could either explicitly `fail`/`failStage` or just allow the exception to be thrown and have akka streams deal with it. I'm not sure I have a strong preference between those two, you?

Reply:
ok, sounds good to me. Then current code does the right thing already.